### DISCUSSION TITLE
How to categorise transactions?

### DISCUSSION THREAD

**Discussion by harshilsharma63** [ananthakumaran/paisa#91] [Q&A]
Is there a way to auto-categorise or maybe semi-automatic way? Can I manually export the ledger file, run a python script on it to categorise stuff for myself and re-import it?

**Comment by ananthakumaran** [comment]
There are 2 ways to categorize

1) When you import transactions from CSV (or other formats), you can use [predictAccount](https://paisa.fyi/reference/import/#predictaccountterms-string-prefix-string-string) function to categorize 
2) You can use [Bulk Edit](https://paisa.fyi/reference/bulk-edit/) to change accounts for all matching transactions.

If this doesn't work for you, you can share more context here, that might help me to come up with other solutions. You can also check https://plaintextaccounting.org/ for more tools.

**Reply by harshilsharma63** [reply]
Checking it out

**Reply by harshilsharma63** [reply]
Okay, so docs for predictAcount (and kudos, they are good docs!) say this-

> Prediction will only work if you have similar transactions in ledger file. It usually means, first time you have to manually fix the Unknown account and then subsequent imports will work

So if I import one full year worth of transactions at once from a single file, it will be better if I break it up into, say, quarterly, import one quarter, manually categorize stuff in there, then import remaining quarters so it can learn from existing entries?

**Reply by ananthakumaran** [reply]
Yup, that would work. I personally import csv from banks/credit cards once every month and put it in a separate file (one per month per bank/cc) etc. In the main file, I would do [include](https://paisa.fyi/reference/journal/#include)

```
include visa/*.ledger
include bank/sbi/*.ledger
```

First import is always a hassle though, since predict account won't work

**Reply by harshilsharma63** [reply]
Makes sense. I have a python script for categorizing my bank and credit card statement. I can just run the same on the generated ledger as its text based as well.

**Reply by harshilsharma63** [reply]
A side question - I see the SBI credit card support is missing. Is it because its a PDF format? Are you open to accepting PRs for adding support for it?

**Reply by ananthakumaran** [reply]
> I see the SBI credit card support is missing

It's because I couldn't get a sample file. Usually, I ask users to share anonymized sample files and include the file in the repo https://github.com/ananthakumaran/paisa/tree/master/fixture/import so it will get tested on changes and can be updated in the future. PDF files are usually hard to anonymize, so not sure I can add them.

**Reply by harshilsharma63** [reply]
Okay, I'll see if I can help.

**Reply by harshilsharma63** [reply]
@ananthakumaran I've created a anonymized SBI credit card statement PDF. Should I go ahead and create a PR for adding it?

**Reply by harshilsharma63** [reply]
I'll also work on parsing this in the app if you think you can't pick this up soon, like this week. I can work on it right now.

**Reply by harshilsharma63** [reply]
I see you don't have a well-documented developer guide. Is there any?

**Reply by ananthakumaran** [reply]
Could you share the pdf file here, I can handle the rest. There is no documentation about the code, all the documentation are aimed at users as I don't have any plans to accept outside code contribution as of now (check the readme).

**Reply by harshilsharma63** [reply]
[sbi.pdf](https://github.com/ananthakumaran/paisa/files/13303579/sbi.pdf)
Okay, here you go.

**Reply by ananthakumaran** [reply]
I have added the template to [master](https://github.com/ananthakumaran/paisa/commit/97fa346981ac67a502fb314a83a13126eeafed75). You would need the master since it uses a new template helper function. Though, I couldn't add the pdf file to import test yet, haven't figured out how to get pdf-js  to work on node js.

**Reply by harshilsharma63** [reply]
Hey, yes, I saw your commit. Trying it out now. I've got master running. Got a dev environment up and running. It was pretty easy actually.

**Reply by harshilsharma63** [reply]
It works perfectly! That was super quick! I tested with statements spanning multiple pages and it worked great there as well! 🚀🚀🚀

So basically we code can open any PDF, convert to text and then we simply apply handlebar parsing on it?

**Reply by ananthakumaran** [reply]
Paisa already had support for pdf, I just added the handlebar template yesterday. 

There are two independent parts, first is the conversion of csv/xlsx/pdf to two dimensional array (like spreadsheet). Once that is done, rest is handled by handlebar template. Converting csv/xlsx to two dimensional array is straightforward, pdf on the other handle is more involved. In the sbi cc case, the cell detection inside the row was not handled properly since the description has uneven spacing etc. So I had to do regexp at the row level to extract data.

**Reply by harshilsharma63** [reply]
Got it. This makes adding new PDF based processing so much easier.

**Reply by harshilsharma63** [reply]
BTW not a SBI card issue, but I see that both ICICI and SBI credit card statements include the transaction about receiving bill amount. Paisa is interpreting this as in income in cashflow and expenses section. While from a ledger POV, it sounds correct, it doesn't sound right if I go by cashflow diagram to understand money in and money out.

**Reply by harshilsharma63** [reply]
For example this ledger entry from CC statement-

```ledger
2023/01/02 NEFT PAYMENT RECEIVED
    Income:Refund:CreditCard:ICICI        -5000.68 INR
    Liabilities:CreditCard:ICICI
```

Because the bank statement will contain a debit entry for this, and we're considering this CC entry as income, the whole sum is probably not correct. I'm no ledger expert so I may be wrong. Thoughts?

**Reply by ananthakumaran** [reply]
This is something that's not handled by the default template. Usually, I delete this transaction (since the bank statement will have the same transaction anyway).

In general, if you import for 2 independent accounts (bank/credit cards etc) and there is an inter-account transfer, bank -> bank or bank -> cc, etc you have to delete it from one account's journal and keep it on the other account's journal. In the CC example you can do it manually or update the handlebar template to ignore this transaction (could be done by matching the description)

**Reply by harshilsharma63** [reply]
Do we have some sort of `ignore` category keyword such that any transaction marked with this category is ignored from all calculations? I checked the docs but couldn't find anything. If we had such a keyword, we could mark such transactions as ignored and future imports will automatically get this category from matching. That way, you don't need to manually delete a bunch of transactions every time you import something.

**Reply by ananthakumaran** [reply]
It's not there as of today, but a similar feature has been requested by another user as well, let me think about it. Paisa shows the verbatim output of ledger cli balance command in editor page, so the ledger cli also has to ignore the transaction to avoid confusion, which I don't know if possible.

**Reply by harshilsharma63** [reply]
Understood.
